### ISSUE TITLE
Evaluation Period 1: Weeks 2 & 3

### ISSUE THREAD

**Issue by coatless** [Yashwants19/RcppMLPACK#4]
Split off from https://github.com/Yashwants19/RcppMLPACK/issues/2#issuecomment-636495013

After completing the above work I will open PRs one-by-one in a branch as suggested by @rcurtin.

 Complete tests (if some tests still lefts)
 Prepare complete directory structure using cmake in mlpack :
-- Replace version in DESCRIPTION using cmake (${MLPACK_VERSION}).
-- Build complete NAMESPACE file using cmake only.
-- Copying all the mlpack files to RcppMLPACK/src/mlpack
-- Generate RcppExports.cpp/.R if possible didn't explore this part yet.
-- Build config file from config.ac
 After the generating the directory, Install/Build the package using R CMD.
 Test the package with handwritten bindings (Initially).

**Comment by rcurtin** [comment]
A question about this one---since most of this work involves setting up the CMake infrastructure for the mlpack repository, does it make more sense to open a PR and review in Yashwants19/mlpack instead of here?  I imagine that will make it easier to make the PR for mlpack when everything is finally ready.  (Sorry to have so many administrative questions and concerns. :))

**Comment by eddelbuettel** [comment]
I think working at the RcppMLPACK repo is easy(er).  

> Build complete NAMESPACE file using cmake only.

Are we sure that is a good idea?  R has generators for this (package `roxygen2`).  We can illustrate.

> Generate RcppExports.cpp/.R if possible didn't explore this part yet.

Same. Rcpp Attributes does that, just invoke `compileAttributes()`.

**Comment by Yashwants19** [comment]
> A question about this one---since most of this work involves setting up the CMake infrastructure for the mlpack repository, does it make more sense to open a PR and review in Yashwants19/mlpack instead of here? I imagine that will make it easier to make the PR for mlpack when everything is finally ready. (Sorry to have so many administrative questions and concerns. :))

Hi I think so, this sound more reasonable to me. As from now onward most of work will perform under `mlpack/src/mlpack/binidings/R`. But for communication we can use these issue tickets, they work best for the communication. For any doubts I can ask you all here only.


> Are we sure that is a good idea? R has generators for this (package roxygen2). We can illustrate.

Yes, we can use this.

> Same. Rcpp Attributes does that, just invoke compileAttributes().

Yeah, my concern was that cmake must invoke `compileAttributes()` for us, after generating all the bindings(`src/.cpp` and `R/.R` files) automatically.

**Comment by eddelbuettel** [comment]
Just have CMake (or anything else) call R:  `Rscript -e "compileAttributes()"`.

There is an additional (important !!) step to find the "right R" because during package tests the `$PATH` may be adjusted to point to another one, or the `RHOME` variable may, so in `autoconf` we often do these check lines first:

```
## Set R_HOME, respecting an environment variable if one is set
: ${R_HOME=$(R RHOME)}
if test -z "${R_HOME}"; then
    AC_MSG_ERROR([Could not determine R_HOME.])
fi
```

and then always call via

```
## Use R to set CXX and CXXFLAGS
CXX=$(${R_HOME}/bin/R CMD config CXX)
CXXFLAGS=$("${R_HOME}/bin/R" CMD config CXXFLAGS)
```

but that is really a later detail for an R package for CRAN -- not for builds within MLPACK.

**Comment by Yashwants19** [comment]
>> Are we sure that is a good idea? R has generators for this (package roxygen2). We can illustrate.

>Yes, we can use this.

Hey, I am getting some problem here. when I invoke:
`Rscript -e "devtools::document()"` or `Rscript -e "roxygen2::roxygenise()"`
It is loading before creating namespace and that too gives error.
```
First time using roxygen2. Upgrading automatically...
Loading mlpackr
Error in Rcpp::compileAttributes(path) : 
  pkgdir must refer to the directory containing an R package
``` 
I mange to resolve this error by creating NAMESPACE with the comment `# Generated by roxygen2: do not edit by hand` it works good but it pre load everything.

**Comment by eddelbuettel** [comment]
My answer is simple.  Don't do `devtools` and other obfuscation tools.

One (very easy) way around / with it is to work from RStudio. It will call these tools, but from the correct directory.

I have a very personal workflow I like which is often command-line drive, and I have a bunch of helper scripts in [this examples directory of the command-line front-end `r`](https://github.com/eddelbuettel/littler/tree/master/inst/examples) -- notably `compAttr.r` to call `Rcpp::compileAttributes()` for you, and `roxy.r` to call `roxygen2::roxygenize()` with the right args.  

But you can approximate either quickly from bash with these which you could also `alias`

     Rscript -e 'Rcpp::compileAttributes()'               # run in the package source dir
     Rscript -e 'roxygen2::roxygenize(".")'               # run in the package source dir

(I didn't double check args now, and I had my helpers for years ....)

**Comment by eddelbuettel** [comment]
And I just did 

```sh
cd ~/git/rcppmlpack-gsoc2020     # my dir of your repo
git co -                         # back to master, co is a git alias for check out
git pu                           # pull with an option or two, maybe --all --prune 
compAttr.r
roxy.r
```

which resulted in

```
Loading required package: Rcpp
Loading required package: xml2
Writing lars.Rd
Writing pca.Rd
Writing test_r_binding.Rd
```

**Comment by rcurtin** [comment]
Awesome, if we can do it all through bash and `Rscript` (or similar) then it should be easy to automate the steps with CMake. :+1:

**Comment by eddelbuettel** [comment]
Yes. R is extremely scriptable.

**Comment by coatless** [comment]
@Yashwants19 I think we need you to write another blog post for week 2. 

How is the `cmake` integration going? Was using `Rscript` approach @eddelbuettel mentioned helpful?

**Comment by Yashwants19** [comment]
Hi @coatless, can I post a combine blog of week 2-3? But If you want I can post a blog for this week too.
I have pushed a commit yesterday [here](https://github.com/Yashwants19/mlpack/commit/796d43056c84bb458065d04ec3368c71c90ab7b9), but I want to make some more changes before creating a PR with full instruction. My work of week 2-3 is almost complete, hence for remaining days of week 3 we can use this for review purpose. 
Thanks a lot @eddelbuettel for `Rscript` approach this works great.
May be I will create a PR later today or may be tomorrow. We can run the initial commit by:
(R, Rcpp, RcppArmadillo and BH must be present on the system. This can detect R but not other dependencies.)
```
git clone https://github.com/Yashwants19/mlpack.git
cd mlpack
git checkout R-directory-structure
mkdir build && cd build
cmake ..
make R                             # For building
ctest -R r_binding_test --verbose  # For testing
```

Thank You.

**Comment by eddelbuettel** [comment]
> R, Rcpp, RcppArmadillo and BH must be present on the system.

Yes. The R package management system can take care of that pretty easily.  You can also ask R from CMake if you wish:

```r
R> pkg <- c("R", "Rcpp", "RcppArmadillo", "BH")
R> IP <- install.packages()
R> IP <- installed.packages()
R> match(pkg, IP[,1])
[1]  NA 471 474  39
R> 
```

R gets a `NA` as it is not a package.

**Comment by rcurtin** [comment]
> Hi @coatless, can I post a combine blog of week 2-3? But If you want I can post a blog for this week too.

I think it would be a good idea to publish a post, even if it's basically saying "this week I am almost done with the week 2/3 goals... stay tuned for a cool update next week". This is for continuity, so that anyone watching the blog (but not this repo) can know what's going on. :) Does that seem reasonable?

**Comment by Yashwants19** [comment]
> Yes. The R package management system can take care of that pretty easily. You can also ask R from CMake if you wish:

Mentors call, I can add this to cmake no problem. But usually mlpack ask user to have all package install before build the bindings from source. Using `findpackage.cmake` mlpack checks whether the package is present or not in system. I think so from RcppMLPACK/mlpackr repo there will be no problem with the installation process.

**Comment by Yashwants19** [comment]
> I think it would be a good idea to publish a post, even if it's basically saying "this week I am almost done with the week 2/3 goals... stay tuned for a cool update next week". This is for continuity, so that anyone watching the blog (but not this repo) can know what's going on. :) Does that seem reasonable?

This sounds reasonable, I will add a blog soon :).

**Comment by eddelbuettel** [comment]
> Mentors call, I can add this to cmake no problem. 

Don't waste time on it. As I said, it is _automatically available_ when you use an R package.

**Comment by Yashwants19** [comment]
> Don't waste time on it. As I said, it is automatically available when you use an R package.

This sounds good :).

**Comment by rcurtin** [comment]
From the mlpack CMake side, we'll definitely need to check that the R dependencies are installed before deciding to build the R bindings.  So we will certainly need a `FindR.cmake` or something just to find R itself (maybe we can adapt something like https://github.com/root-project/root/blob/master/cmake/modules/FindR.cmake ?).  Then, we'll also need to ensure that the R dependencies (`Rcpp`, `RcppArmadillo`, `BH`) are all available, but it sounds like we can just do that by calling R with the script @eddelbuettel suggested.

Keep in mind the whole process here will have to have two logical steps: 

 1. We'll have to use mlpack's build system to generate the R bindings. (All my comment here is addressing that part.)

 2. Once we've generated the R bindings using mlpack's build system, we'll need to make that available as a package in CRAN.

Once step (2) is done, then a user can just install mlpack via CRAN and it should be smooth sailing from there.  Maybe it's superfluous to say that, so sorry if I am preaching to the choir. :) Hopefully it will mostly only be us and some build systems that ever compile mlpack's R bindings with, e.g., `cmake -DBUILD_R_BINDINGS=ON ..` or similar. :+1:

**Comment by Yashwants19** [comment]
Hello everyone, here is the [link](https://medium.com/@yashwantsingh.sngh/gsoc-week-2-3-part-1-fbf7c1f00c4a) for my weekly blog. Sorry but I am late here. Kindly share your views.
Thank You.

**Comment by coatless** [comment]
@Yashwants19 we'll need Week 3's Blog to go live soon as well.

**Comment by Yashwants19** [comment]
Closed by [#6](https://github.com/Yashwants19/mlpack/pull/6).
